File: Practica 8/HerenciaMult.java
class A {
    public int x;
    public int z;

    public A(int x, int z) {
        this.x = x;
        this.z = z;
    }

    public void incrementaXZ() {
        this.x = this.x + 1;
        this.z = this.z + 1;
    }

    public void incrementaZ() {
        this.z = this.z + 1;
    }
}

class B {
    public int y;
    public int z;

    public B(int y, int z) {
        this.y = y;
        this.z = z;
    }

    public void incrementaYZ() {
        this.y = this.y + 1;
        this.z = this.z + 1;
    }

    public void incrementaZ() {
        this.z = this.z + 1;
    }
}

class D extends A {
    B b;

    public D(int x, int y, int z) {
        super(x, z);
        this.b = new B(y, z);
    }

    public void incrementaXYZ() {
        this.x = this.x + 1;
        this.b.y = this.b.y + 1;
        this.z = this.z + 1;
        this.b.z = this.b.z + 1;
    }

    public void incrementaYZ() {
        this.b.incrementaYZ();
    }

    // Java no permite herencia múltiple, así que debemos manejar B manualmente
    public void incrementaZB() {
        this.b.incrementaZ();
    }

    @Override
    public String toString() {
        return "(x=" + this.x + ", y=" + this.b.y + ", z=" + this.z + ")";
    }
}

public class HerenciaMult {
    public static void main(String[] args) {
        D d = new D(5, 10, 15);
        System.out.println("original: " + d);
        System.out.println("\nUsando los métodos de A:");
        d.incrementaXZ();
        System.out.println("Después de incrementaXZ: " + d);
        d.incrementaZ();
        System.out.println("Después de incrementaZ: " + d);
        System.out.println("\nUsando los métodos de B:");
        d.incrementaYZ();
        System.out.println("Después de incrementaYZ: " + d);
        d.incrementaZB(); // Nota: En Java necesitamos un método diferente para B
        System.out.println("Después de incrementaZ (de B): " + d);
        System.out.println("\nUsando el método de D:");
        d.incrementaXYZ();
        System.out.println("Después de incrementaXYZ: " + d);
    }
}

File: Practica 8/HerenciaMult.py
class A:
    def __init__(self, x, z):
        self.x = x
        self.z = z
    def incrementaXZ(self):
        self.x = self.x + 1
        self.z = self.z + 1
    def incrementaZ(self):
        self.z = self.z + 1
class B:
    def __init__(self, y, z):
        self.y = y
        self.z = z
    def incrementaYZ(self):
        self.y = self.y + 1
        self.z = self.z + 1    
    def incrementaZ(self):
        self.z = self.z + 1
class D(A, B):
    def __init__(self, x, y, z):
        A.__init__(self, x, z)
        B.__init__(self, y, z)    
    def incrementaXYZ(self):
        self.x = self.x + 1
        self.y = self.y + 1
        self.z = self.z + 1    
    def __str__(self):
        return f"(x={self.x}, y={self.y}, z={self.z})"
class Main:
    d = D(5, 10, 15)
    print(f"original: {d}")
    print("\nUsando los métodos de A:")
    d.incrementaXZ()
    print(f"Después de incrementaXZ: {d}")
    d.incrementaZ()
    print(f"Después de incrementaZ: {d}")
    print("\nUsando los métodos de B:")
    d.incrementaYZ()
    print(f"Después de incrementaYZ: {d}")
    d.incrementaZ()
    print(f"Después de incrementaZ: {d}")
    print("\nUsando el método de D:")
    d.incrementaXYZ()
    print(f"Después de incrementaXYZ: {d}")